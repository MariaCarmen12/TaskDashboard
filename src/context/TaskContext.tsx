import { createSlice, configureStore, PayloadAction } from '@reduxjs/toolkit';
import { Task } from '../interfaces/taskTypes';

interface TaskState {
  tasks: Task[];
  filteredTasks: Task[];
}

const initialState: TaskState = {
  tasks: [],
  filteredTasks: [],
};

const taskSlice = createSlice({
  name: 'tasks',
  initialState,
  reducers: {
    addTask: (state, action: PayloadAction<Task>) => {
      state.tasks.push(action.payload);
      state.filteredTasks = state.tasks;
    },
    editTask: (state, action: PayloadAction<Task>) => {
      const index = state.tasks.findIndex(task => task.id === action.payload.id);
      const indexaux = state.filteredTasks.findIndex(task => task.id === action.payload.id);
      if (index !== -1 && indexaux !== -1) {
        state.tasks[index] = action.payload;
        state.filteredTasks[indexaux] = action.payload;
      }
    },
    deleteTask: (state, action: PayloadAction<string>) => {
      state.tasks = state.tasks.filter(task => task.id !== action.payload);
      state.filteredTasks = state.tasks;
    },
    toggleComplete: (state, action: PayloadAction<string>) => {
      const index = state.filteredTasks.findIndex(task => task.id === action.payload);
      const indexaux = state.tasks.findIndex(task => task.id === action.payload);
      if (index !== -1 && indexaux !== -1) {
        state.tasks[indexaux].completed = !state.tasks[indexaux].completed;
        state.filteredTasks[index].completed = !state.filteredTasks[index].completed;
      }
    },
    sortTasks: (state, action: PayloadAction<string>) => {
      if (action.payload === 'Date') {
        state.filteredTasks = state.tasks.slice().sort((a, b) => {
          const dateA = new Date(a.createdAt);
          const dateB = new Date(b.createdAt);
          return dateB.getTime() - dateA.getTime();
        });
      } else if (action.payload === 'Priority') {
        const priorityOrder = { 'High': 1, 'Medium': 2, 'Low': 3 };
        state.filteredTasks = state.tasks.slice().sort((a, b) => priorityOrder[a.priority] - priorityOrder[b.priority]);
      }
    },
    filterTasksByPriority: (state, action: PayloadAction<string>) => {
      const priority = action.payload;
      state.filteredTasks = state.tasks.filter(task => priority === 'All' || task.priority === priority);
    },
    reorderTasks: (state, action: PayloadAction<Task[]>) => {
      const reorderedTasks = action.payload;
      state.filteredTasks = reorderedTasks;
      state.tasks = reorderedTasks;
      state.tasks = state.tasks.map(task => reorderedTasks.find(reorderedTask => reorderedTask.id === task.id) || task);
    },
    cleanTask: (state) => {
      state.filteredTasks = [];
      state.tasks = [];
    }
  },
});

export const { addTask, editTask, deleteTask, toggleComplete, sortTasks, filterTasksByPriority, reorderTasks, cleanTask } = taskSlice.actions;

const store = configureStore({
  reducer: {
    tasks: taskSlice.reducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export default store;
