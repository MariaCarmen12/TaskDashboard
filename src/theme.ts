import { createTheme } from '@mui/material/styles';

// Definición de breakpoints
const breakpoints = {
  values: {
    xs: 0,
    sm: 600,
    md: 900,
    lg: 1200,
    xl: 1536,
  },
};

// Tema claro
const lightTheme = createTheme({
  palette: {
    mode: 'light',
    primary: {
      main: '#1976d2', // Azul principal
      contrastText: '#ffffff', // Texto en contraste para botones primarios
    },
    secondary: {
      main: '#9c27b0', // Morado secundario
      light: "#F1F1F1"
    },
    background: {
      default: '#f4f6f8', // Color de fondo principal
      paper: '#ffffff',   // Color de fondo para elementos como tarjetas
    },
    text: {
      primary: '#333333', // Color de texto principal
      secondary: '#666666', // Color de texto secundario
    },
    error: {
      main: '#d32f2f', // Rojo para errores
    },
    warning: {
      main: '#ffa000', // Amarillo para advertencias
    },
    info: {
      main: '#0288d1', // Azul para información
    },
    success: {
      main: '#2e7d32', // Verde para éxito
    },
  },
  breakpoints,
});

// Tema oscuro
const darkTheme = createTheme({
  palette: {
    mode: 'dark',
    primary: {
      main: '#90caf9', // Azul claro para modo oscuro
      contrastText: '#000000', // Texto en contraste para botones primarios
    },
    secondary: {
      main: '#ce93d8', // Morado claro para modo oscuro
      light: "#1E1E1E"
    },
    background: {
      default: '#121212', // Fondo oscuro principal
      paper: '#1e1e1e',   // Fondo para elementos como tarjetas
    },
    text: {
      primary: '#ffffff', // Texto claro en modo oscuro
      secondary: '#b0bec5', // Texto secundario en modo oscuro
    },
    error: {
      main: '#f44336', // Rojo para errores
    },
    warning: {
      main: '#ffb74d', // Amarillo para advertencias
    },
    info: {
      main: '#29b6f6', // Azul para información
    },
    success: {
      main: '#66bb6a', // Verde para éxito
    },
  },
  breakpoints,
});

export { lightTheme, darkTheme };
