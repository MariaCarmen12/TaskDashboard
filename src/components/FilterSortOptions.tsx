import React, { useState, useEffect } from 'react';
import {  MenuItem,  Box } from '@mui/material';
import { useDispatch } from 'react-redux';
import { sortTasks, filterTasksByPriority, addTask, editTask } from '../context/TaskContext';
import { useTheme } from '@mui/material/styles';

import { Container, FilterIcon, FilterTitle, StyledTextField } from './Styles/FilterSortOptionsStyles';

const FilterSortOptions: React.FC = () => {
  const dispatch = useDispatch();
  const theme = useTheme();

  const [sortOption, setSortOption] = useState<'Date' | 'Priority'>('Date');
  const [filterOption, setFilterOption] = useState<'All' | 'High' | 'Medium' | 'Low'>('All');

  
  useEffect(() => {
    dispatch(filterTasksByPriority(filterOption));
    dispatch(sortTasks(sortOption));
  }, [addTask, editTask, filterOption, sortOption, dispatch]);

 
  const handleSortChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newSortOption = e.target.value as 'Date' | 'Priority';
    setSortOption(newSortOption);
    dispatch(sortTasks(newSortOption));
  };

  const handleFilterChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newFilterOption = e.target.value as 'All' | 'High' | 'Medium' | 'Low';
    setFilterOption(newFilterOption);
    dispatch(filterTasksByPriority(newFilterOption));
  };

  return (
    <Container theme={theme}>
      <Box display="flex" alignItems="center">
        <FilterIcon theme={theme} />
        <FilterTitle variant="h6" theme={theme}>
          Filters
        </FilterTitle>
      </Box>

      <StyledTextField
        select
        label="Sort by"
        value={sortOption}
        onChange={handleSortChange}
        theme={theme}
      >
        <MenuItem value="Date">Date</MenuItem>
        <MenuItem value="Priority">Priority</MenuItem>
      </StyledTextField>

      <StyledTextField
        select
        label="Filter by Priority"
        value={filterOption}
        onChange={handleFilterChange}
        theme={theme}
      >
        <MenuItem value="All">All</MenuItem>
        <MenuItem value="High">High</MenuItem>
        <MenuItem value="Medium">Medium</MenuItem>
        <MenuItem value="Low">Low</MenuItem>
      </StyledTextField>
    </Container>
  );
};

export default FilterSortOptions;


