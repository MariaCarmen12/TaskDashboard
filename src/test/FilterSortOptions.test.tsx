import React from 'react';
import { render, screen, fireEvent, within } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ThemeProvider } from '@mui/material/styles';
import configureStore from 'redux-mock-store';
import FilterSortOptions from '../components/FilterSortOptions';
import { lightTheme } from '../theme'; 
import { RootState } from '../context/TaskContext';

const mockStore = configureStore([]);
const initialState: RootState = {
  tasks: {
    tasks: [],
    filteredTasks: [],
  },
};
let store:any;

const mockTasks = [
  {
    id: '1',
    title: 'Test Task 1',
    description: 'Description 1',
    priority: 'High',
    completed: false,
  },
  {
    id: '2',
    title: 'Test Task 2',
    description: 'Description 2',
    priority: 'Medium',
    completed: false,
  },
];

beforeEach(() => {
  store = mockStore({
    ...initialState,
    tasks: {
      tasks: mockTasks,
      filteredTasks: mockTasks,
    },
  });
});

describe('FilterSortOptions Component', () => {
  const setup = () =>
    render(
      <Provider store={store}>
        <ThemeProvider theme={lightTheme}>
          <FilterSortOptions />
        </ThemeProvider>
      </Provider>
    );

  it('renders correctly', () => {
    setup();
    expect(screen.getByText('Filters')).toBeInTheDocument();
    expect(screen.getByLabelText('Sort by')).toBeInTheDocument();
    expect(screen.getByLabelText('Filter by Priority')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  it('changes sort option', () => {
    setup();
    const filterSelect = screen.getAllByRole('combobox');
    fireEvent.mouseDown(filterSelect[0]);
    const listbox = screen.getAllByRole('option')
    fireEvent.click(listbox[1]);
    expect(filterSelect[0]).toHaveTextContent('Priority');
    // Further test if the correct action was dispatched if needed
  });

  it('changes filter option', () => {
    setup();
    const filterSelect = screen.getAllByRole('combobox');
    fireEvent.mouseDown(filterSelect[1]);
    const listbox = screen.getAllByRole('option')
    fireEvent.click(listbox[1]);
    expect(filterSelect[1]).toHaveTextContent('High');
    // Further test if the correct action was dispatched if needed
  });
});
