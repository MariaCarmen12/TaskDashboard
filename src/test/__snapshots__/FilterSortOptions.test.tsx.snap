// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSortOptions Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-gyf1yz"
  >
    <div
      class="MuiBox-root css-70qvj9"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-qucq6i-MuiSvgIcon-root"
        data-testid="FilterListIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 18h4v-2h-4zM3 6v2h18V6zm3 7h12v-2H6z"
        />
      </svg>
      <h6
        class="MuiTypography-root MuiTypography-h6 css-1i2ud1d-MuiTypography-root"
      >
        Filters
      </h6>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root css-14q8ftn-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-17dx2ji-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r4:"
        id=":r4:-label"
      >
        Sort by
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-20gy86-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          aria-controls=":r5:"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby=":r4:-label :r4:"
          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-10l5tdt-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r4:"
          role="combobox"
          tabindex="0"
        >
          Date
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          tabindex="-1"
          value="Date"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Sort by
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root css-14q8ftn-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-17dx2ji-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r6:"
        id=":r6:-label"
      >
        Filter by Priority
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-20gy86-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          aria-controls=":r7:"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby=":r6:-label :r6:"
          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-10l5tdt-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r6:"
          role="combobox"
          tabindex="0"
        >
          All
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          tabindex="-1"
          value="All"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Filter by Priority
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
